File: app/src/main/java/com/fyp/agrifarm/app/weather/ui/WeatherFragment.java
package com.fyp.agrifarm.app.weather.ui;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.RecyclerView;

import com.fyp.agrifarm.R;
import com.fyp.agrifarm.app.weather.model.HourlyObject;
import com.fyp.agrifarm.app.weather.model.WeatherSharedViewModel;

import java.util.List;

public class WeatherFragment extends Fragment {


    public static final String TAG = "WeatherFragment";
//    private WeatherViewModel weatherViewModel;
     private WeatherSharedViewModel weatherModel;

    private String temperatue;


    public WeatherFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        ConstraintLayout parent = (ConstraintLayout) inflater.inflate(R.layout.content_weather, container,
                false);

//        parent.findViewById(R.id.layout_rel).setLayoutDirection(View.LAYOUT_DIRECTION_RTL);
//        weatherViewModel = new ViewModelProvider(requireActivity()).get(WeatherViewModel.class);
        RecyclerView rvHourlyForecast = parent.findViewById(R.id.rvHourlyForecast);
        WeatherHourlyRecyclerAdapter hourlyRecyclerAdapter = new WeatherHourlyRecyclerAdapter(getContext());
//        weatherViewModel.getHourlyforcastlist().observe(getViewLifecycleOwner(), hourlyRecyclerAdapter::updateList);
//        weatherSharedViewModel = new ViewModelProvider(this).get(WeatherSharedViewModel.class);
//        weatherSharedViewModel.init();
//        weatherSharedViewModel.gethourlyList().observe(getViewLifecycleOwner(),hourlyRecyclerAdapter::changeDataSource);
        weatherModel = new ViewModelProvider(this).get(WeatherSharedViewModel.class);
        weatherModel.init();
//        weatherModel.gethourlyList().observe(getViewLifecycleOwner(), new Observer<List<HourlyObject>>() {
//            @Override
//            public void onChanged(List<HourlyObject> list) {
//                Log.d("LISTWALA"," " +  list.size()) ;
//                hourlyRecyclerAdapter.changeDataSource(list);
//            }
//        });
        weatherModel.gethourlyList().observe(getViewLifecycleOwner(),hourlyRecyclerAdapter::updateList);
        rvHourlyForecast.setAdapter(hourlyRecyclerAdapter);

        RecyclerView rvDailyForecast = parent.findViewById(R.id.rvDailyForecast);
        WeatherDailyRecyclerAdapter dailyRecyclerAdapter = new WeatherDailyRecyclerAdapter(getContext());
        weatherModel.getDailyList().observe(getViewLifecycleOwner(), dailyRecyclerAdapter::updateList);
        rvDailyForecast.setAdapter(dailyRecyclerAdapter);


        TextView wftemperatute = parent.findViewById(R.id.wftvWeatherTemp);
        TextView wfdescription = parent.findViewById(R.id.wftvweatherdescription);
        TextView wfwindpressure = parent.findViewById(R.id.wftvWeatherWind);
        TextView wfhumidity = parent.findViewById(R.id.wftvWeatherHumidity);
        TextView wfday = parent.findViewById(R.id.wftvWeatherDay);
        ImageView weatherIcon = parent.findViewById(R.id.wfivWeatherIcon);

        weatherModel.getCurrentWeather().observe(getViewLifecycleOwner(), weatherDailyForecast -> {

            SharedPreferences sharedPref = PreferenceManager
                    .getDefaultSharedPreferences(requireContext());
            String WeatherPref = sharedPref
                    .getString("weatherUnit", "-1");
            if (WeatherPref.equals("Celsius"))
            {
                temperatue = weatherDailyForecast.getTemperature() + "°C";

            }
            else
            {
                temperatue = weatherDailyForecast.getTemperature() + "°F";

            }

            wftemperatute.setText(temperatue);
            wfdescription.setText(weatherDailyForecast.getDescription());
            wfday.setText(weatherDailyForecast.getDay());
            wfhumidity.setText(weatherDailyForecast.getHumidity().substring(0,2));
            wfwindpressure.setText(weatherDailyForecast.getWindPressure());

            // Wrapped with catch incase resource ID not found
            try {
                String id = weatherDailyForecast.getIconurl();
                weatherIcon.setImageResource(weatherModel.weatherIconsMap.get(id));
            } catch (Exception e) {
                e.printStackTrace();
            }

        });


        return parent;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
    }
}

File: app/src/main/java/com/fyp/agrifarm/app/weather/ui/WeatherFragment.kt
package com.fyp.agrifarm.app.weather.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.RecyclerView
import com.fyp.agrifarm.R
import com.fyp.agrifarm.app.toastFrag
import com.fyp.agrifarm.app.weather.model.CELSIUS
import kotlinx.android.synthetic.main.content_weather.*
import kotlinx.android.synthetic.main.content_weather.view.*
import java.util.*


class WeatherFragment : Fragment() {
    //    private WeatherViewModel weatherViewModel;
    private lateinit var weatherModel: WeatherSharedViewModel


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val parent = inflater.inflate(R.layout.content_weather, container,
                false) as ConstraintLayout


        val rvHourlyForecast: RecyclerView = parent.findViewById(R.id.rvHourlyForecast)
        val hourlyRecyclerAdapter = WeatherHourlyRecyclerAdapter(context)

        weatherModel = ViewModelProvider(this).get(WeatherSharedViewModel::class.java)
        weatherModel.updateWeather(null)

        weatherModel.hourlyList.observe(viewLifecycleOwner, Observer {
            hourlyRecyclerAdapter.updateList(it)
        })

        rvHourlyForecast.adapter = hourlyRecyclerAdapter

        val rvDailyForecast: RecyclerView = parent.findViewById(R.id.rvDailyForecast)
        val dailyRecyclerAdapter = WeatherDailyRecyclerAdapter(context)

        weatherModel.dailyList.observe(viewLifecycleOwner, Observer {
            dailyRecyclerAdapter.updateList(it)
        })
        rvDailyForecast.adapter = dailyRecyclerAdapter


        var isNight = false
        val c = Calendar.getInstance()
        val timeOfDay = c[Calendar.HOUR_OF_DAY]

        if (timeOfDay in 0..5) {
            isNight = true
        } else if (timeOfDay in 6..18) {
            isNight = false
        } else if (timeOfDay in 19..23) {
            isNight = true
        }

        val res = if (isNight) R.mipmap.bg_weather_night else R.mipmap.bg_weather_day
        parent.bgWeather.setImageResource(res)

        val wftemperatute = parent.findViewById<TextView>(R.id.wftvWeatherTemp)
        val wfdescription = parent.findViewById<TextView>(R.id.wfWeatherDescription)
        val wfwindpressure = parent.findViewById<TextView>(R.id.wftvWeatherWind)
        val wfhumidity = parent.findViewById<TextView>(R.id.wftvWeatherHumidity)
        val weatherIcon = parent.findViewById<ImageView>(R.id.wfivWeatherIcon)


        weatherModel.currentWeather.observe(viewLifecycleOwner, Observer {
            val sharedPref = PreferenceManager.getDefaultSharedPreferences(requireContext())
            val weatherPref = sharedPref.getString("weatherUnit", "-1")
            val temperature = if (weatherPref == CELSIUS) {
                it.temperature + "°C"
            } else {
                it.temperature + "°F"
            }
            wftemperatute.text = temperature

            wfdescription.text = it.description
            wfhumidity.text = it.humidity

            // Wrapped with catch incase resource ID not found

            try {
                it.icon?.let {  ic ->
                    weatherIcon.setImageResource(ic)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        })

        return parent
    }

    companion object {
        const val TAG = "WeatherFragment"
    }
}